Fixes for Fails with shared VPC being specified using a reference #47

ingwarr opened the pull request:
Fixes https://github.com/terraform-google-modules/terraform-google-project-factory/issues/47

ingwarr commented:
@aaron-lane Conflicts resolved I think we need to add a new variable `shared_vpc_subnets_count` to README.md and examples

kopachevsky commented on modules/core_project_factory/main.tf:
```diff
@@ -172,8 +172,8 @@ resource "google_project_services" "project_services_authority" {
-resource "google_compute_shared_vpc_service_project" "shared_vpc_attachment" {
-  count = var.shared_vpc != "" ? 1 : 0
+resource "google_compute_shared_vpc_service_project" "shared_vpc_attachment"
+  count = var.shared_vpc_subnets_count != null ? 1 : 0
```
does it make sense to create boolean variable for `var.shared_vpc_subnets_count != null` ?

ingwarr replied:
Yes it makes sense, I'll try to test it

kopachevsky commented on modules/core_project_factory/main.tf:
```diff
@@ -283,7 +283,7 @@ resource "google_service_account_iam_member" "service_account_grant_to_group" {
  *****************************************************************************************************************/
 resource "google_project_iam_member" "controlling_group_vpc_membership" {
-  count = var.shared_vpc != "" && length(compact(var.shared_vpc_subnets)) == 0 ? local.shared_vpc_users_length : 0
+  count = var.shared_vpc_subnets_count != null && length(compact(var.shared_vpc_subnets)) == 0 ? local.shared_vpc_users_length : 0
```
same for length(compact(var.shared_vpc_subnets)) , we can define value of this expression it as a local variale and just use it?

ingwarr replied:
Better use the boolean variable from your previous comment if it will work

ingwarr commented:
> I'm not sure this is the correct solution to the issue. The error may be resolved because we've removed `var.shared_vpc` from the counts, but what is the benefit of adding `var.shared_vpc_subnets_count` when the counts already use `length(compact(var.shared_vpc_subnets))`?
> 
> Since the `core_project_factory` submodule is private, we have some flexibility to introduce breaking changes to that submodule without affecting the root module. What if we replace `var.shared_vpc` with a boolean variable in the counts, and set that variable based on the contents of `var.shared_vpc` in the root module? This change would also be necessary in the `gsuite_enabled` submodule.
> 
> Example of `main.tf` in the root module:
> 
> ```diff
> module "project-factory" {
>   source = "./modules/core_project_factory"
> 
>   group_email                 = module.gsuite_group.email
>   group_role                  = var.group_role
>   lien                        = var.lien
>   manage_group                = var.group_name != "" ? "true" : "false"
>   random_project_id           = var.random_project_id
>   org_id                      = var.org_id
>   name                        = var.name
>   project_id                  = var.project_id
>   shared_vpc                  = var.shared_vpc
> + shared_vpc_enabled          = var.shared_vpc != ""
>   billing_account             = var.billing_account
>   folder_id                   = var.folder_id
>   sa_role                     = var.sa_role
>   activate_apis               = var.activate_apis
>   apis_authority              = var.apis_authority
>   usage_bucket_name           = var.usage_bucket_name
>   usage_bucket_prefix         = var.usage_bucket_prefix
>   credentials_path            = var.credentials_path
>   shared_vpc_subnets          = var.shared_vpc_subnets
>   labels                      = var.labels
>   bucket_project              = var.bucket_project
>   bucket_name                 = var.bucket_name
>   bucket_location             = var.bucket_location
>   auto_create_network         = var.auto_create_network
>   disable_services_on_destroy = var.disable_services_on_destroy
>   default_service_account     = var.default_service_account
>   disable_dependent_services  = var.disable_dependent_services
> }
> ```

@aaron-lane  I've implemented this suggestion and it works, wonderful, thank you!!!

ingwarr commented:
@aaron-lane I think I need to rebase it

aaron-lane commented:
There are multiple failures within the project-factory-shared-vpc control of the full-local instance: https://concourse.infra.cft.tips/builds/5646#L5d1c47b5:647 (search for "expected" to find them quickly).

ingwarr commented:
> There are multiple failures within the project-factory-shared-vpc control of the full-local instance: https://concourse.infra.cft.tips/builds/5646#L5d1c47b5:647 (search for "expected" to find them quickly).

@aaron-lane I don't have appropriate privileges to run the full test locally Error: [ERROR] Error creating group: googleapi: Error 403: Not Authorized to access this resource/api, forbidden

        on ../../../modules/gsuite_enabled/main.tf line 46, in resource "gsuite_group" "group":
        46: resource "gsuite_group" "group" {

bohdanyurov-gl commented:
@ingwarr Issue is probably related to this:
https://github.com/terraform-google-modules/terraform-google-project-factory/blob/2d181abb0da9b8dc4b8090ba3504764488ce9a6f/modules/gsuite_enabled/variables.tf#L67
https://github.com/terraform-google-modules/terraform-google-project-factory/blob/2d181abb0da9b8dc4b8090ba3504764488ce9a6f/modules/gsuite_enabled/main.tf#L47

Looks like a regression. Please also correctly rebase your changes.

aaron-lane commented on modules/core_project_factory/main.tf:
```diff
@@ -443,7 +440,7 @@ resource "google_storage_bucket_iam_member" "api_s_account_storage_admin_on_proj
   provider = google-beta
 
-  count = local.gke_shared_vpc_enabled && length(compact(var.shared_vpc_subnets)) != 0 ? length(var.shared_vpc_subnets) : 0
+  count = var.shared_vpc_enabled && length(compact(var.shared_vpc_subnets)) != 0 ? length(var.shared_vpc_subnets) : 0
```
This logic should not be changed. The definition of `local.gke_shared_vpc_enabled` must be updated to use `var.shared_vpc_enabled`.

ingwarr replied:
Fixed

aaron-lane commented on modules/core_project_factory/main.tf:
```diff
@@ -470,8 +467,7 @@ resource "google_compute_subnetwork_iam_member" "gke_shared_vpc_subnets" {
 resource "google_project_iam_member" "gke_host_agent" {
-  count = local.gke_shared_vpc_enabled ? 1 : 0
-
+  count = var.shared_vpc_enabled ? 1 : 0
```
Same issue about `local.gke_shared_vpc_enabled`.

ingwarr replied:
Fixed

aaron-lane commented on modules/core_project_factory/variables.tf:
```diff
@@ -147,3 +147,8 @@ variable "disable_dependent_services" {
+variable "shared_vpc_enabled" {
+  description = "Existing of shared vpc"
+  type        = bool
+  default     = false
```
This should not have a default. We want to ensure that both public modules explicitly pass this argument.

ingwarr replied:
Done

aaron-lane commented on variables.tf:
```diff
@@ -148,3 +148,9 @@ variable "disable_dependent_services" {
   type        = string
 }
 
+variable "shared_vpc_enabled" {
```
This variable should not be defined here.

ingwarr replied:
Removed

bohdanyurov-gl commented on modules/core_project_factory/main.tf:
```diff
@@ -43,7 +43,7 @@ locals {
   )
   api_s_account_fmt      = format("serviceAccount:%s", local.api_s_account)
-  gke_shared_vpc_enabled = var.shared_vpc != "" && contains(var.activate_apis, "container.googleapis.com") ? "true" : "false"
+  gke_shared_vpc_enabled = var.shared_vpc_enabled && contains(var.activate_apis, "container.googleapis.com") ? true : false
```
you don't need ternary operator here

ingwarr replied:
Fixed

ingwarr replied:
Fixed

bohdanyurov-gl commented on modules/core_project_factory/variables.tf:
```diff
@@ -143,7 +147,13 @@ variable "default_service_account" {
+}
+
+variable "shared_vpc_enabled" {
+  description = "Existing of shared vpc"
```
"If shared VPC should be used"

ingwarr replied:
Done

bohdanyurov-gl commented on variables.tf:
```diff
@@ -16,7 +16,8 @@
 variable "random_project_id" {
   description = "Enables project random id generation. Mutually exclusive with project_id being non-empty."
-  default     = "false"
+  type = bool
```
formatting

bohdanyurov-gl replied:
Not related to your change, please revert

ingwarr replied:
Reverted

bohdanyurov-gl commented on variables.tf:
```diff
@@ -121,19 +123,20 @@ variable "bucket_location" {
   description = "Whether project services will be disabled when the resources are destroyed"
-  default     = "true"
-  type        = string
+  default     = true
```
Not related to your change, please revert

ingwarr replied:
ok

bohdanyurov-gl commented on variables.tf:
```diff
@@ -121,19 +123,20 @@ variable "bucket_location" {
   description = "Add a lien on the project to prevent accidental deletion"
-  default     = "false"
-  type        = string
+  default     = false
```
Not related to your change, please revert

ingwarr replied:
Reverted

bohdanyurov-gl commented on variables.tf:
```diff
@@ -121,19 +123,20 @@ variable "bucket_location" {
 variable "auto_create_network" {
   description = "Create the default network"
-  default     = "false"
+  type = bool
```
Not related to your change, please revert

ingwarr replied:
Reverted

bohdanyurov-gl commented on variables.tf:
```diff
@@ -68,7 +69,8 @@ variable "sa_role" {
 
 variable "apis_authority" {
   description = "Toggles authoritative management of project services."
-  default     = "false"
```
Not related to your change, please revert

ingwarr replied:
reverted

bohdanyurov-gl commented on variables.tf:
```diff
@@ -144,7 +147,12 @@ variable "default_service_account" {
 variable "disable_dependent_services" {
   description = "Whether services that are enabled and which depend on this service should also be disabled when this service is destroyed."
-  default     = "true"
-  type        = string
```
Not related to your change, please revert

ingwarr replied:
reverted

bohdanyurov-gl commented on variables.tf:
```diff
@@ -144,7 +147,12 @@ variable "default_service_account" {
 }
 
+variable "shared_vpc_enabled" {
+  description = "Existing of shared vpc"
```
Better description

ingwarr replied:
Done

bohdanyurov-gl commented on docs/upgrading_to_project_factory_v1.0.md:
```diff
@@ -18,12 +18,12 @@ while the "project-factory-gsuite" module does.
 /// @file main.tf
 
 provider "google" {
-  credentials = "${file(var.credentials_path)}"
```
Please remove all changes except shared_vpc configs

ingwarr replied:
ok

ingwarr replied:
Done

bohdanyurov-gl commented:
@aaron-lane changes that you've proposed wouldn't work (actually, @ingwarr has already verified that they don't). The primary reason is that your variable still depends on computed value, so terraform can not build full graph on plan stage.

I've suggested @ingwarr to implement separate variable "shared_vpc_enabled", that doesn't rely on any computed resources at all.

Please note, that while this is out of context of the issue, I know for sure that the same problem is going to happen with shared_vpc_subnets (they are not covered in tests and issue create don't use them). @ingwarr please open a separate issue to fix subnets as well.

ingwarr commented:
![image](https://user-images.githubusercontent.com/2316547/61964153-e6684a00-afd5-11e9-84b7-889316138a43.png)
![image](https://user-images.githubusercontent.com/2316547/61964191-fa13b080-afd5-11e9-8504-19a5fb5355c3.png)
![image](https://user-images.githubusercontent.com/2316547/61964229-0c8dea00-afd6-11e9-861f-52906bc64a7c.png)
![image](https://user-images.githubusercontent.com/2316547/61964262-1e6f8d00-afd6-11e9-815f-cda0e4e11a46.png)
![image](https://user-images.githubusercontent.com/2316547/61964295-36dfa780-afd6-11e9-9e93-b7a6b7c3620e.png)
![image](https://user-images.githubusercontent.com/2316547/61964332-4b23a480-afd6-11e9-9009-a347ccfa0aca.png)
